etcdserver/api/v3rpc/watch.go

```diff
@@ -61,6 +68,10 @@ type serverWatchStream struct {
 	watchStream storage.WatchStream
 	ctrlStream  chan *pb.WatchResponse
 
+	// progress tracks the watchID that stream might need to send
+	// progress to.
+	progress map[storage.WatchID]bool
```

Comment:
if there is a recent event sent to a watch, we can skip it when sending progress.